multicluster/controllers/multicluster/member/serviceexport_controller.go

```diff
@@ -626,10 +665,106 @@ func (r *ServiceExportReconciler) updateOrCreateResourceExport(resName string,
 	return nil
 }
 
+// getSubsetsFromEndpointSlice will check all ready Endpoints from any matched EndpointSlice
+// The purpose is to simplify Endpoints change detection since we already mapped the EndpointSlice
+// change event into a ServiceExport event. It will be hard to track individual EndpointSlices change
+// after mapping. We may change ServivceExport controller into a traditional controller without controller
+// runtime framework in the future if we need to track individual EndpointSlice change.
+func (r *ServiceExportReconciler) getSubsetsFromEndpointSlice(ctx context.Context, req ctrl.Request) ([]corev1.EndpointSubset, bool, error) {
+	epSliceList := &discovery.EndpointSliceList{}
+	hasReadyEndpoints := false
+	err := r.Client.List(ctx, epSliceList, &client.ListOptions{
+		LabelSelector: getEndpointSliceLabelSelector(req.Name),
+		Namespace:     req.Namespace})
+	if err != nil {
+		return nil, hasReadyEndpoints, err
+	}
+	if len(epSliceList.Items) == 0 {
+		return nil, hasReadyEndpoints, nil
+	}
+	var subsets []corev1.EndpointSubset
+	for _, eps := range epSliceList.Items {
+		if eps.AddressType == discovery.AddressTypeIPv4 {
+			ports := convertPorts(eps.Ports)
+			for _, ep := range eps.Endpoints {
+				if *ep.Conditions.Ready {
+					// We only cares if there is ready Endpoints for a Service when the endpointIPType is ClusterIP,
+					// so skip handling the EndpointSubset and stop the loop early if any ready address is found.
+					if r.endpointIPType == "ClusterIP" {
+						return nil, hasReadyEndpoints, nil
+					}
+					subset := corev1.EndpointSubset{}
+					readyAddresses := ipToAddresses(ep.Addresses)
+					subset.Addresses = append(subset.Addresses, readyAddresses...)
+					subset.Ports = ports
+					subsets = append(subsets, subset)
+				}
+			}
+		}
+	}
+	return subsets, len(subsets) > 0, nil
+}
+
+func (r *ServiceExportReconciler) getSubsetsFromEndpoint(ctx context.Context, req ctrl.Request, eps *corev1.Endpoints) (bool, error) {
+	hasReadyEndpoints := false
+	err := r.Client.Get(ctx, req.NamespacedName, eps)
+	epIsNotFound := apierrors.IsNotFound(err)
+	if err == nil {
+		for _, subsets := range eps.Subsets {
+			if len(subsets.Addresses) > 0 {
+				return true, nil
+			}
+		}
+		return hasReadyEndpoints, nil
+	}
+	if !epIsNotFound {
```

Comment:
I feel the following is easier to understand:

```
if apierrors.IsNotFound(err) {
        return false, nil
}
klog.ErrorS(err, "Failed to get Endpoints", req.String())
return false , err
```

Reply:
If the error is logged in the caller, then no need to log here, but append the err message to the returned error.

Reply:
Moved the log in the caller.